Fix error deleting unicode char in entry widget

lucor opened the pull request:
This PR fixes a slice bounds out of range error when a unicode char is deleted using "Backspace" in the entry widget. (gl driver)

Additionally updates the entry widget to use internally runes (aiming to help with #38)
This has introduced a BC removing the entry.Text field in favour of entry.Text() method.

lucor commented on widget/entry.go:
```diff
@@ -161,10 +162,15 @@ func (e *Entry) Hide() {
+}
 
+// updateRunes updates the internal text to the given value
+func (e *Entry) updateRunes(runes []rune) {
```
Please consider this name like a placeholder. Should we rename into something more self explaining? 
Example:
- `e.runes` -> `e.content`
- `updateRunes` -> `updateContent`

lucor replied:
renamed to updateText into 439ed1ff8a31a2eaf50ce7e43b8267043c63bb54

andydotxyz commented:
This is very helpful thanks.
However the removal of the Text field means that developers can no longer do this:
&widget.Entry{Text:"Default"}
I am trying to support both constructor function style and this direct style of widget and canvas object creation...

What do you think?

lucor commented:
I see, developer can still do
```
e := &widget.Entry{}
e.SetText("Default")
```
it is more verbose but the alternative is to expose the runes buffer like:
`&widget.Entry{Text:[]rune("Default")}`
(renaming back e.runes to e.Text)

Note: if we want follow this approach we'd also make the password field public.

Wrt constructor I'm wondering if makes sense to a NewMultiline too (it could help with internal logic: i.e when newline must be ignored)
```
// NewEntry creates a new entry widget.
func NewEntry(text string) *Entry {
	e := &Entry{}
        e.SetText(text)
	Renderer(e).Layout(e.MinSize())
	return e
}

// NewPasswordEntry creates a new entry password widget
func NewPasswordEntry(passoword string) *Entry {
	e := &Entry{password: true}
        e.SetText(passoword)
	Renderer(e).Layout(e.MinSize())
	return e
}

// NewMultilineEntry creates a new multiline entry widget
func NewMultilineEntry(text string) *Entry {
	e := &Entry{multiline: true}
        e.SetText(text)
	Renderer(e).Layout(e.MinSize())
	return e
}
```

andydotxyz commented:
Yes I think it makes sense to expose Password and Multiline - I don't have a strong feeling about the multiline constructor but happy if you think it's a good idea.

You're right exposing the Text as runes is kindof clunky looking. Is it sensible to have SetText() be string, but the text buffer be rune? Is it a big overhead to store as string instead of rune?

As always I think it's important that this API be as easy to use as possible, and I think that includes exposing basic state as public fields. As well as being a user facing API decision it also plays into the ticket #5 I think...

lucor commented:
I've updated the code to rollback the Text field and removed the Text() method.
The only impact I can see is the conversion from string to runes and back into insertToCursor function but probably for now is good enough.

With regard to the NewMultilineEntry, I'll create another PR for it.

andydotxyz commented:
Great, thanks

andydotxyz commented:
I added the Multiline field in commit [b274b2](https://github.com/fyne-io/fyne/commit/b274b28a11963e8a82c086808003c5146ab109ac).
I updated the Password field to be public as well.

Apologies if this overrode some local changes you had, but this made a lot of sense so I wanted it in :)

lucor commented:
No worries, it tooks some time since I was trying to have also the label widget with unicode support introducing a basic text widget. I hope to have something ready to share for discussion soon :)
